### ISSUE TITLE
New feature request: LXC container

### ISSUE THREAD

**Issue by davxiao** [photoprism/photoprism#147]
Is there going to be LXC container build for photoprism?

**Comment by lastzero** [comment]
Not yet, need to finish the frontend first...

**Comment by iGadget** [comment]
Having just migrated all of my infra to Proxmox, this would be a *very* welcome addition. Is there any way I can assist to make this happen?

**Comment by timdonovanuk** [comment]
Just a thought, but an LXC container would make accessing hardware like GPU (for tensorflow/machine learning/facial recognition and video transcoding) way way way easier!

**Comment by TapuGithub** [comment]
Also using proxmox. One additional thing will be possible in case this feature will be implemented - sharing video transcoding accelerator(QSV in my case) with other containers(Plex in my case).

**Comment by lastzero** [comment]
Using host devices, e.g. for machine learning or transcoding, is also possible with Docker. Why is it easier with LXC? Don't have any experience with that :/

**Comment by TapuGithub** [comment]
> Using host devices, e.g. for machine learning or transcoding, is also possible with Docker. Why is it easier with LXC? Don't have any experience with that :/

I think using host devices is almost identical in Proxmox(LXC) or with Docker. But Proxmox does much more than only containers - VMs, backups, clustering, etc.  And there is no docker containers support in Proxmox - only LXC or VM with Docker in it. That's why we want "native" to Proxmox way - LXC.

**Comment by timdonovanuk** [comment]
I'm not entirely sure, but passing through GPU hardware host -> VM -> docker -> process feels like a challenge! In theory with LXC it just goes host -> process (or thereabouts).

**Comment by TapuGithub** [comment]
> I'm not entirely sure, but passing through GPU hardware host -> VM -> docker -> process feels like a challenge! In theory with LXC it just goes host -> process (or thereabouts).

Yes - it's much trickier. Also, you can only pass GPU to one VM, but you can pass it to multiple containers. I'm already using my integrated intel GPU to transcode video in Plex, and it would be great if I could use it in Photoprism to occasionally transcode video.

**Comment by timdonovanuk** [comment]
> you can only pass GPU to one VM, but you can pass it to multiple containers.

I never even thought of that! Great point.

**Comment by grassel82** [comment]
Photoprism works in LXC containers in Proxmox if you install docker inside LXC container. You just need to enable `keyctl` and `nesting` inside your `OPTIONS`, if not `docker-compose` will fail with error messages. I use this configuration and it is quite fast. Storage is an M2 SSD disk, so speed is no issue.

Kind regards,

Gilles

**Comment by TapuGithub** [comment]
Well, using the app in os-level virtualization in a virtual machine in hypervisor seems overly complicated for me. 
Also, the main reasons for using 'native' LXC are to reduce virtualization overhead and allow direct access to the host's devices from multiple containers(like GPU).

**Comment by Qix-** [comment]
@TapuGithub **IssueHunt is abandoned.** If you are putting money into it, you are losing it. Nobody will be able to receive that money as IssueHunt is not responding to withdrawal requests.

**Comment by timdonovanuk** [comment]
Just to shed some insight or practical experience on this, I recently got the Frigate NVR docker image working in an LXC container. This was preferable to having it running in a VM running docker, because it does image recognition and passing through hardware acceleration ("/dev/dri/renderD128") is key to having it run. It's quite possible I could pass through the hardware to a VM, and then pass through that to docker, but people have had mixed reports of this working.

**Comment by tteck** [comment]
PhotoPrism LXC script can be found here https://tteck.github.io/Proxmox/

**Comment by timdonovanuk** [comment]
Holy crap @tteck that is an incredible list of invaluable resources. Bookmarked! You should share this on reddit (/r/selfhosted)!

**Comment by jojo-gfx** [comment]
Seems it doesn't work correct?

Breaks with a failure on "Dependencies" I would think (that with the warning of Patience),  and then VM is running but have no root login and web-frontend refuses the connection. :(
In advanced mode now I have a root login, but web console isn't available so I can't do anything now.
Was looking so great first, but so it's a bit bad... 

Here the output (also was try default and also advanced:

Creating a PhotoPrism LXC using the above advanced settings
 ✓ Using local for Template Storage.
 ✓ Using SSD1TB for Container Storage.
 ✓ Updated LXC Template List
 ✓ LXC Container 102 was successfully created.
 ✓ Started LXC Container
 ✓ Set up Container OS
 ✓ Network Connected: 192. ... deleted
 ✓ Internet Connected
 ✓ DNS Resolved github.com to 140.82.121.3
 ✓ Updated Container OS
 ✓ Installed Dependencies
 ✓ Set up Node.js Repository
 ✓ Installed Node.js
 - Installing Golang (Patience)...‼ ERROR 1@104 Unknown failure occurred.

A try out of update in the container itself is quit with:

Stopping PhotoPrism... Failed to stop photoprism.service: Unit photoprism.service not loaded.

So the software isn't running.


... try to make a new make from GIT but seems, the LXC data is really really old and nothing is passing so right. Many errors and many deprecated infos. :(

**Comment by ricketyboo** [comment]
@jojo-gfx I just recently did a photoprism lxc ct build based on these instructions and it's been working great. I am using Ubuntu instead of debian but not much difference there really. Maybe it'll help with your issues 
https://github.com/IndrekHaav/photoprism-debian

**Comment by harshavmb** [comment]
> @jojo-gfx I just recently did a photoprism lxc ct build based on these instructions and it's been working great. I am using Ubuntu instead of debian but not much difference there really. Maybe it'll help with your issues https://github.com/IndrekHaav/photoprism-debian

While this is great! I have a question about migrations/upgrades. You would need to make some schema changes right. 
Currently, the docker-compose handles this. Is this something you have considered?
I'm also very biased about running this directly on a VM rather than on a container. Going to give it a try soon..

**Comment by jojo-gfx** [comment]
Hi, after many hours I have now a good way also to install it in a VM. :)
I also using Ubuntu (in version 22.04) and if I do install docker directly in the setup I have no problems anymore.
Also the port settings in yml-file were a mistake but after I understand the line my problems were solved. :)
But, if you have a good designed container, please make a upload for other users, because this is a really good way for not-docker-server-professionals ... the most matching users are to find in this category! :)

**Comment by timdonovanuk** [comment]
FYI there is a Photoprism LXC helper script available here: https://tteck.github.io/Proxmox/

It has a wizard that talks you through setting up the LXC and even has an automatic update script available.

I would highly recommend it + the other LXC templates on that page. 

tl;dr: run `bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/photoprism-v4.sh)"` on your Proxmox host and you'll get a Photoprism LXC.

**Comment by jojo-gfx** [comment]
Yes, but the script is running into Errors after a while of installation on proxmox (most usecase)... (seems it is too old?).
Or is it a new script now?

**Comment by timdonovanuk** [comment]
There was an update on 2022-08-19, I thought that fixed the issues. Not tested though! There are currently no open issues regarding it so I think it's now fine.

**Comment by jojo-gfx** [comment]
My test was later so the script still has problems, could be with Proxmox 7 only?

**Comment by jojo-gfx** [comment]
Oh, I've test it now again and nor it works! :D 
Last period I've test it the script was hanging up on dependencies and was frozen, as far I know.
Could be there were any ressources not available? Would be nice the sript would not fail there if any goes wrong there or stop installation after short time and give a error like "try it later"?
So it seems if all ressources are available it works also on Proxmox 7! :D

**Comment by jaydhary14** [comment]
> Oh, I've test it now again and nor it works! :D Last period I've test it the script was hanging up on dependencies and was frozen, as far I know. Could be there were any ressources not available? Would be nice the sript would not fail there if any goes wrong there or stop installation after short time and give a error like "try it later"? So it seems if all ressources are available it works also on Proxmox 7! :D

How to upgrade, not able to find upgrade script

**Comment by jojo-gfx** [comment]
What do you mean? Ugrading you have to do it in the running machine itself.

But after one or two weeks everytime one of the internal Docker containers is stopped and I have to push it to run ... :(

**Comment by lastzero** [comment]
@tteck @davxiao @timdonovanuk We now provide *tar.gz* installation packages at [**dl.photoprism.app/pkg/linux/**](https://dl.photoprism.app/pkg/linux/). These can be used to install PhotoPrism in a [LXC base image](https://images.linuxcontainers.org/) of your choice, without building it from source:

- https://dl.photoprism.app/pkg/linux/
- https://dl.photoprism.app/pkg/linux/README.html

You can, for example, run the following commands to download and install PhotoPrism in `/opt/photoprism`:

```
sudo mkdir -p /opt/photoprism
cd /opt/photoprism
wget -c https://dl.photoprism.app/pkg/linux/amd64.tar.gz -O - | sudo tar -xz
sudo ln -sf /opt/photoprism/bin/photoprism /usr/local/bin/photoprism
photoprism --version
```

Since many people seem to be using the following Proxmox / LXC installation script, we'd be grateful if you could help us update it to use our *tar.gz* packages so that development dependencies are no longer required (the maintainer seems to accept pull requests):

- https://github.com/tteck/Proxmox/blob/main/install/photoprism-install.sh


Related Issues / PRs:

- https://github.com/IndrekHaav/photoprism-debian/issues/22
- https://github.com/IndrekHaav/photoprism-debian/pull/23

*Note that the minimum required glibc version is 2.35, so for example Ubuntu 22.04 and Debian Bookworm will work with these binaries, but older Linux distributions may not be compatible. Also note that the build ID does not match our Docker images at the moment. However, they will match after our next stable Docker release.*

**Comment by lastzero** [comment]
I've also added a small *libheif* installation guide to our download server so you can install the latest `heif-convert` binary along with PhotoPrism (required for HEIC support):

- https://dl.photoprism.app/dist/libheif/
- https://dl.photoprism.app/dist/libheif/README.html

**Comment by lastzero** [comment]
Today I learned that you can convert our Docker image to the LXC format e.g. with the following commands:

```bash
apt update
apt install lxc umoci skopeo
lxc-create photoprism -t oci -- --url docker://photoprism/photoprism:latest
```

This didn't come as a total surprise to me since Docker and LXC are pretty much the same technology. Is there a reason why none of our LXC users have ever tried or mentioned this? :thinking: 

PhotoPrism can then be configured and started like any other LXC container:

```bash
lxc-start --name=photoprism -- /opt/photoprism/bin/photoprism start
```

If an experienced LXC user contributes information on the commands/steps required to complete the network, storage and database configuration, we will be happy to pass it on to less experienced users, e.g. in the LXC section of our Getting Started FAQ:
- https://docs.photoprism.app/getting-started/faq/#lxc-images

**Comment by mintsoft** [comment]
> This didn't come as a total surprise to me since Docker and LXC are pretty much the same technology. Is there a reason why none of our LXC users have ever tried or mentioned this? 🤔
> 
> PhotoPrism can then be configured and started like any other LXC container:

So having looked at the resulting container, this doesn't really work well enough to be useful. The created container doesn't have enough of the underlying systems (it is missing any and all ip related packages and doesn't have a way to be able to install them without .deb-dropping into the container, which isn't really a valid method), so I'm not sure we can say that it's a viable method. 

That said, I think this issue is actually about creating an LXC template for Proxmox rather than an LXC container itself; LXC differs from docker in that it is much more like a VM (in that it is persistent etc) and is usually managed as such (i.e. get a debian container and install a .deb or whatever the mechanism is for the given software). 

Looking at the proxmox wiki, it seems like the list of LXC templates in Proxmox actually comes from turnkeylinux (ref:https://pve.proxmox.com/wiki/Linux_Container) so presumably however _that_ works, is how we can add Photoprism in.

**Comment by lastzero** [comment]
Thanks for your feedback! I had it working as it is using the commands shown in our FAQ. Which additional packages should be installed in the container image for it to be usable for you? How about we simply pre-install them in an upcoming release (unless there any other issues that you believe cannot be solved)?

**Comment by mintsoft** [comment]
> Which additional packages should be installed in the container image for it to be usable for you?

https://packages.debian.org/bookworm/iproute2 and https://packages.debian.org/bookworm/isc-dhcp-client are the two that stand out to be able to manage the IPs and get one via DHCP. I don't know if you install those into the docker container if that'll be a happy situation though; will that be OK from a docker perspective?

Even if we do get the conversion working, I don't think that will actually address the above requests as they are asking for it be available from there and there isn't a supported way (that I know of anyway!) to import LXC templates from anywhere other than the list in the UI (I'll check this when I'm able to, I might have overlooked something) .

**Comment by lastzero** [comment]
Since Ubuntu is based on Debian, the same packages (but in a newer version) can be installed in our latest Docker base image:

- https://packages.ubuntu.com/oracular/isc-dhcp-client
- https://packages.ubuntu.com/oracular/iproute2

Since I already have too many other issues on my plate, I unfortunately won't be able to experiment with LXC on my own (I also don't know enough about how users will deploy and use it in practice). However, I'll be happy to add them so you and others can test it!?

**Comment by mintsoft** [comment]
> Since Ubuntu is based on Debian, the same packages (but in a newer version) can be installed in our latest Docker base image:

Yes sorry, Ubuntu if its an Ubuntu image. Force of habit makes me use packages.debian.org but that will work.

>Since I already have too many other issues on my plate, I unfortunately won't be able to experiment with LXC on my own (I also don't know enough about how users will deploy and use it in practice). However, I'll be happy to add them so you and others can test it!?

If it's a quick thing to do then I'd say throw them in and I'll give it a spin.

I've just checked in the Proxmox UI and there is an option to get a container from a URL (I've always just ignored that!) 
![Screenshot_20241207_164135](https://github.com/user-attachments/assets/8419617a-eacf-4d59-88a1-820000fada70)
Not sure what format it's expecting etc but that at-least means there is a way to add them (using `docker://photoprism/photoprism:latest` as the url doesn't work as the UI regex rejects it).

Edit: it looks like this doesn't expect a template (i.e. a shell script), it's actually looking for a `rootfs.tar.xz`

**Comment by lastzero** [comment]
I found a related post on Reddit that looks like it is technically possible, but since "Proxmox is built for LXC", the configuration in `/var/lib/lxd` is "hardcoded":

- https://www.reddit.com/r/Proxmox/comments/18poplg/comment/keq26s7/

This sounds a bit like they either don't have the resources to build a proper integration, or have decided not to support it for whatever reason?

Note that OCI image support is part of the official LXC distribution, so it shouldn't be unusual or problematic:

- https://github.com/lxc/lxc/blob/main/templates/lxc-oci.in

**Comment by mintsoft** [comment]
> This sounds a bit like they either don't have the resources to build a proper integration, or have decided not to support it for whatever reason?
>
>Note that OCI image support is part of the official LXC distribution, so it shouldn't be unusual or problematic:
>
 >   https://github.com/lxc/lxc/blob/main/templates/lxc-oci.in

Yeah I think it's primarily because it (proxmox) is built from an ESXi/baremetal hypervisor perspective rather than like a NAS so it's more about whole machine virtualisation. The Proxmox UI is absolutely going to get in the way of using the OCI system built into LXC as there just aren't the buttons for it. 

I think it should be pretty trivial to build a "convert docker:latest to lxc root.tar.xz"-inator (which is what lxc-oci does) which could run as part of whatever build infrastructure exists etc. If you've got the appetite for that I can spike something out?

**Comment by mintsoft** [comment]
This has certainly nerd-sniped me a smidge, something like this will switch the latest docker into an LXC template that will upload into proxmox and you'll be able to create a container from that etc. . All the executables are there etc, it's just like it's missing the service.

It will get an IP via dhcp if you configure it in proxmox etc (that already works, just very difficult to see without the ip tooling). However it doesn't actually start photoprism at all, presumably something gets lost in the conversion as to what is supposed to happen on startup. All the executables are there etc, it's just like it's missing the service.

```
skopeo copy docker://photoprism/photoprism:latest oci:photoprism-docker-latest:photoprismlxc
targetManifest=$(cat photoprism-docker-latest/index.json | jq '.manifests[0].digest')
mkdir lxctmp
umoci unpack --rootless --image photoprism-docker-latest:photoprismlxc lxctmp
cd lxctmp/rootfs
tar cvfJ ../../photoprism-lxc.tar.xz ./*
cd ../..
rm -Rf lxctmp
# photoprism-lxc.tar.xz is the container for proxmox 
```

I'll have a mooch around and see if anything obvious as to why appears in a bit.

**Comment by mintsoft** [comment]
Ah this is pretty obvious when you see it. The LXC oci template will introduce the config `lxc.execute.cmd` to run photoprism when you use the commandline, however through proxmox it will write out the LXC configuration (https://github.com/proxmox/pve-container/blob/a39c5b0617e15aabf5259f9743337beb16c2c173/src/PVE/LXC.pm#L618 hence the comments on the link above about the hardcoded configuration) and there isn't a way for an LXC template to provide additional configuration to Proxmox (that I can see anyway). 

The only way to make that work in a way that isn't absolute hackery of the hypervisor is to introduce into the container a systemd unit which runs photoprism on startup, which is pretty easy to do and would work. 

I feel like the best option (rather than converting the docker container) might be to build a fresh LXC container and use a binary tar install within it (this depends on what build infrastructure exists I guess). I think I saw that done somewhere for photoprism but I can't find it now and all the documentation only mentions docker installs, if I find it, I'll give that a spin.
